pybamm/solvers/base_solver.py

```diff
@@ -397,287 +449,225 @@ def jacp(*args, **kwargs):
-                            expr = symbol.right
-                            found_t = True
-                        # Dimensional
-                        elif symbol.right.id == (pybamm.t * model.timescale_eval).id:
-                            expr = symbol.left / symbol.right.right
-                            found_t = True
-                        elif symbol.left.id == (pybamm.t * model.timescale_eval).id:
-                            expr = symbol.right / symbol.left.right
-                            found_t = True
-
-                        # Update the events if the heaviside function depended on t
-                        if found_t:
+            model.y0 = initial_conditions(0, y_zero, inputs)
+
+        if ics_only:
+            pybamm.logger.info("Finish solver set-up")
+            return
+
+        # Check for heaviside and modulo functions in rhs and algebraic and add
+        # discontinuity events if these exist.
+        # Note: only checks for the case of t < X, t <= X, X < t, or X <= t,
+        # but also accounts for the fact that t might be dimensional
+        # Only do this for DAE models as ODE models can deal with discontinuities
+        # fine
+        if len(model.algebraic) > 0:
+            for symbol in itertools.chain(
+                model.concatenated_rhs.pre_order(),
+                model.concatenated_algebraic.pre_order(),
+            ):
+                if isinstance(symbol, _Heaviside):
+                    found_t = False
+                    # Dimensionless
+                    if symbol.right.id == pybamm.t.id:
+                        expr = symbol.left
+                        found_t = True
+                    elif symbol.left.id == pybamm.t.id:
+                        expr = symbol.right
+                        found_t = True
+                    # Dimensional
+                    elif symbol.right.id == (pybamm.t * model.timescale_eval).id:
+                        expr = symbol.left / symbol.right.right
+                        found_t = True
+                    elif symbol.left.id == (pybamm.t * model.timescale_eval).id:
+                        expr = symbol.right / symbol.left.right
+                        found_t = True
+
+                    # Update the events if the heaviside function depended on t
+                    if found_t:
+                        model.events.append(
+                            pybamm.Event(
+                                str(symbol), expr, pybamm.EventType.DISCONTINUITY,
+                            )
+                        )
+                elif isinstance(symbol, pybamm.Modulo):
+                    found_t = False
+                    # Dimensionless
+                    if symbol.left.id == pybamm.t.id:
+                        expr = symbol.right
+                        found_t = True
+                    # Dimensional
+                    elif symbol.left.id == (pybamm.t * model.timescale_eval).id:
```

Comment:
@tinosulzer: I'm not understanding this line. In line 150-153 we ensure that model.timescale_eval is a scalar, so when you do `pybamm.t * model.timescale_eval` wont this be simplified to `pybamm.t` so this elif is the same as the if above it?

Reply:
`pybamm.t` is just an object that returns `t` when `.evaluate(t=t)` is called (not a constant). So when multiplied by a scalar it doesn't get simplified

Reply:
But I'm happy to merge this as is. Those lines that aren't being covered weren't covered before already

Reply:
no worries, I want to cover those lines now!

Reply:
think I know what I was doing wrong now.....